File: internal/go/api/ollama.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type OllamaClient struct {
	BaseURL string
	HTTP    *http.Client
}

func NewOllamaClient(baseURL string) *OllamaClient {
	return &OllamaClient{
		BaseURL: baseURL,
		HTTP: &http.Client{
			Timeout: 300 * time.Second,
		},
	}
}

type GenerateRequest struct {
	Model  string   `json:"model"`
	Prompt string   `json:"prompt"`
	Stream bool     `json:"stream"`
	Images []string `json:"images,omitempty"`
	Format string   `json:"format,omitempty"`
}

type GenerateResponse struct {
	Model    string `json:"model"`
	Response string `json:"response"`
	Done     bool   `json:"done"`
}

func (c *OllamaClient) Generate(req GenerateRequest) (*GenerateResponse, error) {
	url := fmt.Sprintf("%s/api/generate", c.BaseURL)
	payload, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	res, err := c.HTTP.Post(url, "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(res.Body)
		return nil, fmt.Errorf("ollama error (status %d): %s", res.StatusCode, string(body))
	}

	var resp GenerateResponse
	if err := json.NewDecoder(res.Body).Decode(&resp); err != nil {
		return nil, err
	}

	return &resp, nil
}

func (c *OllamaClient) ListModels() ([]string, error) {
	url := fmt.Sprintf("%s/api/tags", c.BaseURL)
	res, err := c.HTTP.Get(url)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	var data struct {
		Models []struct {
			Name string `json:"name"`
		} `json:"models"`
	}

	if err := json.NewDecoder(res.Body).Decode(&data); err != nil {
		return nil, err
	}

	names := make([]string, len(data.Models))
	for i, m := range data.Models {
		names[i] = m.Name
	}
	return names, nil
}

File: internal/node/services/ollama.ts
import { LLMProvider, ChatMessage, LLMConfig } from '../types';
import { ConfigLoader } from './config-loader';

export class OllamaProvider implements LLMProvider {
    async chat(messages: ChatMessage[], config: LLMConfig): Promise<string> {
        let baseUrl = config.baseUrl;
        if (!baseUrl) {
            const appConfig = await ConfigLoader.load();
            baseUrl = appConfig.ollama.api_url || 'http://127.0.0.1:11434';
        }
        const url = `${baseUrl}/api/chat`;
        try {
            const response = await fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    model: config.model,
                    messages: messages,
                    stream: true, // Use streaming to prevent timeout
                    format: config.format,
                }),
            });

            if (!response.ok) {
                const errorData = await response.json().catch(() => ({ error: response.statusText }));
                throw new Error(`Ollama error [${response.status}]: ${errorData.error || response.statusText}`);
            }

            if (!response.body) {
                throw new Error('No response body received from Ollama');
            }

            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            let fullContent = '';
            let buffer = '';

            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                const chunk = decoder.decode(value, { stream: true });
                buffer += chunk;

                const lines = buffer.split('\n');
                // Keep the last partial line in the buffer
                buffer = lines.pop() || '';

                for (const line of lines) {
                    if (!line.trim()) continue;
                    try {
                        const json = JSON.parse(line);
                        if (json.message && json.message.content) {
                            fullContent += json.message.content;
                        }
                        if (json.done) return fullContent; // Use return instead of break to exit early
                    } catch (e) {
                        console.error('Failed to parse line as JSON:', line);
                    }
                }
            }

            // Process any remaining data in buffer if it exists (though usually done=true means no more data)
            if (buffer.trim()) {
                try {
                    const json = JSON.parse(buffer);
                    if (json.message && json.message.content) {
                        fullContent += json.message.content;
                    }
                } catch (e) {
                    // Ignore, might be truly incomplete
                }
            }

            return fullContent;
        } catch (error: any) {
            console.error('Ollama Chat Fetch Error:', {
                url,
                model: config.model,
                messageCount: messages.length,
                error: error.message,
                cause: error.cause
            });
            throw new Error(`Failed to communicate with Ollama (${config.model} @ ${url}): ${error.message}${error.cause ? ' (Cause: ' + error.cause + ')' : ''}`);
        }
    }

    async listModels(baseUrl?: string): Promise<string[]> {
        let effectiveUrl = baseUrl;
        if (!effectiveUrl) {
            const appConfig = await ConfigLoader.load();
            effectiveUrl = appConfig.ollama.api_url || 'http://127.0.0.1:11434';
        }

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 5000); // 5s timeout

        try {
            const url = `${effectiveUrl}/api/tags`;
            console.log(`Fetching models from: ${url}`);
            const response = await fetch(url, { signal: controller.signal });
            clearTimeout(timeoutId);

            if (!response.ok) {
                throw new Error(`Ollama error: ${response.statusText} (${response.status})`);
            }
            const data = await response.json();
            return data.models.map((m: any) => m.name);
        } catch (error: any) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new Error('Connection timeout - is Ollama running?');
            }
            throw new Error(`Failed to connect to Ollama: ${error.message}`);
        }
    }
}
